EIPS/eip-7212.md

```diff
@@ -0,0 +1,180 @@
+
+## Motivation
+
+“secp256r1” elliptic curve is a standardized curve by NIST which has the same calculations by different input parameters with “secp256k1” elliptic curve used by the “ecrecover” precompiled contract. The cost of combined attacks and the security conditions are almost the same for both curves. Adding a precompiled contract which is similar to "ecrecover" can provide signature verifications using the “secp256r1” elliptic curve in the smart contracts and multi-faceted benefits can occur. One important factor is that this curve is widely used and supported in many modern devices such as Apple’s Secure Enclave, Webauthn, Android Keychain which proves the user adoption. Additionally, the introduction of this precompile could enable valuable features in the account abstraction which allows more efficient and flexible management of accounts by transaction signs in mobile devices. 
+Most of the modern devices and applications rely on the “secp256r1” elliptic curve. The addition of this precompiled contract enables the verification of device native transaction signing mechanisms. For example:
+
+1. **Apple's Secure Enclave:** There is a separate “Trusted Execution Environment” in Apple hardware which can sign arbitrary messages and can only be accessed by biometric identification.
+2. **Webauthn:** Web Authentication (WebAuthn) is a web standard published by the World Wide Web Consortium (W3C). WebAuthn aims to standardize an interface for authenticating users to web-based applications and services using public-key cryptography. It is being used by almost all of the modern web browsers.
+3. **Android Keystore:** Android Keystore is an API that manages the private keys and signing methods. The private keys are not processed while using Keystore as the applications’ signing method. Also, it can be done in the “Trusted Execution Environment” in the microchip.
+4. **Passkeys:** Passkeys is utilizing FIDO Alliance and W3C standards. It replaces passwords with cryptographic key-pairs which is also can be used for the elliptic curve cryptography.
+
+Modern devices have these signing mechanisms that are designed to be more secure and they are able to sign transaction data, but none of the current wallets are utilizing these signing mechanisms. So, these secure signing methods can be enabled by the proposed precompiled contract to initiate the transactions natively from the devices and also, can be used for the key management. This proposal aims to reach maximum security and convenience for the key management.
+
+## Specification
+
+As of `FORK_TIMESTAMP` in the integrated EVM chain, add precompiled contract `P256VERIFY` for signature verifications in the “secp256r1” elliptic curve at address `PRECOMPILED_ADDRESS`.
```

Comment:
**Curious if people are interested in `CREATE2` shadowing.**

In short, if we make `PRECOMPILED_ADDRESS` the CREATE2 address of a contract that implements identical functionality--instead of a special low address like `0x00..0019`--then it allows a smoother deployment.

Chains will adopt this EIP at different times; maybe certain L2s are first, others later, and mainnet last. With CREATE2 shadowing, contracts can work on all chains immediately. They just become more gas-efficient once a given chain adds the precompile. A related benefit is that it reduces the need for upgradeability in user contracts.

This is especially salient for this particular precompile, since a primary use-case for it is hardware backed wallets. A 4337 wallet is itself a CREATE2 contract works on all chains with identical code.

Reply:
I like your idea. I see the biggest merit of what you describe being a route reduce barrier for adding precompile contracts, mitigating the chicken-and-egg problem of "needs to demonstrate usage and interest to convince adding precompile" vs "needs adding the precompile to get more usage".

Reply:
The idea looks very interesting. I also would love to follow future plans and integrations about it.

Reply:
@xinbenlv happy to, but does it need an EIP? this is more a general pattern that can be applied to all precompiles that are pure functions

it raises the question of "why haven't we done it so far", but i think there were good reasons:
- early precompiles like [eip 152](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-152.md) didn't use `CREATE2` shadowing because CREATE2 didn't exist yet, and also because cross-chain compatibility wasn't a concern yet
- many L2 precompiles like Optimism's [`0x42....` series](https://www.evmdiff.com/diff?base=1&target=10) can't use it because they're not pure functions (instead, it's a built-in L2 specific oracle or bridge)

for future pure functions precompiles, CREATE2 shadowing seems like a win/win

Reply:
I agree with you having a CREATE2 will be a wthat it's win/win

To answer your question "does it need an EIP"? I'd assume the question being: Do we need to standardize it?

Here is what I think:

When there are many ways and variables ("moving pieces") to do the same thing, standardization helps people coordinate. An example would be screw driver and screw, the size of screw bing standardize so people can manufacture screw and drivers independently and they will work together.

In the CREATE2 shadowing, I presume there are a few "moving pieces." that if people make different choices, they will end up with the different sizes of screw, hence there are value to standardize.

1. For the same functionality, ideally everyone use the same contact address generated by CREATE, so that ultimately, with a hard-fork we can change them to precompile and with lower gas cost. We don't want people generating different contacts deployed at different addresses.

2. How to determine the input of CREATE2? Which `sender`, which `nonce`, which `bytecode`?  
 - 2.1 Are there going to be a sender that everyone want to use so if nonce and bytecode are also determined it we can all agree upon the contract address generated by CREATE2?
 - 2.2 How do we generate `nonce`? Shall we do something like ERC-600 / ERC-601 or BIP-32?
 - 2.3 Should there be [any magic](https://twitter.com/_SamWilsn_/status/1666793849585369088?s=20) in bytecode? E.g. Should bytecode be consistent should there be any additional behavior that caller can expect?
 
 These are a few things I can think of.

To this point, I think the discussion of CREAT2 shadwing seems a bit off topic from this EIP. I'd suggest we make a ethereum-magician.org post to discuss, and leave this PR for the Precomiple of secp256r1 Curve support.

Reply:
Yeah, fair points. The CREATE2 address would eventually have to be part of the proposed precompile regardless. I agree re: standardizing how it's created.

Happy to make an ethmagicians post

Reply:
Considering that the solutions that are expected to be presented as precompiled contracts cannot always be implemented as a Solidity smart contract, can we say that this is a downside for the standardization purpose?

Reply:
Thread moved to ethmagicians: https://ethereum-magicians.org/t/create2-shadowing-for-precompiles/14821/3